scripts/checkpatch.pl

```diff
@@ -4255,12 +4255,11 @@ sub process {
 			     "Prefer dev_$level(... to dev_printk(KERN_$orig, ...\n" . $herecurr);
 		}
 
-# ENOSYS means "bad syscall nr" and nothing else.  This will have a small
-# number of false positives, but assembly files are not checked, so at
-# least the arch entry code will not trigger this warning.
+# Zephyr uses ENOTSUP for cases where ENOTSUP or ENOSYS would be more appropriate;
+# See: https://github.com/zephyrproject-rtos/zephyr/issues/23727
 		if ($line =~ /\bENOSYS\b/) {
 			WARN("ENOSYS",
-			     "ENOSYS means 'invalid syscall nr' and nothing else\n" . $herecurr);
+			     "Use ENOTSUP instead of ENOSYS in Zephyr API\n" . $herecurr);
```

Comment:
will this issue a warning on any mention of ENOSYS or only on function returns?

Reply:
It will diagnose the same uses it does today, but with a message that is meaningful for Zephyr.

Reply:
this means that if I am checking some return code from a module that returns NOSYS  we will have the above warning, which is not what we want:

code like this:
```
ret = some_third_party_lib();
if (ret == -ENOSYS) {
}
```

returns:

```
-:9: WARNING:ENOSYS: ENOSYS means 'invalid syscall nr' and nothing else
#9: FILE: drivers/dma/dma_dw.c:262:
+	if (ret == -ENOSYS) {
```

is this what we want?

Reply:
or with this PR...

```
-:9: WARNING:ENOSYS: Use ENOTSUP instead of ENOSYS in Zephyr API
#9: FILE: drivers/dma/dma_dw.c:262:
+	if (ret == -ENOSYS) {

```

Reply:
The existing message appears to be emitted on every introduction of `ENOSYS` regardless of context, so yes, with this change we would **continue** to get false positives, but with a more relevant message that would make more clear why it should be ignored ("this use isn't in Zephyr API, so we can ignore it").

Historically such false positives don't appear to be a concern.  `dma_dw.c` does not actually have the code that evokes a diagnostic, and in fact are no checks of a return value against `-ENOSYS` in Zephyr today.  There are many checks against `-ENOTSUP`, in drivers, samples, and subsystems as well as tests.

Improving the existing diagnostic is the solution I'm offering.  There's no obligation to accept it.  If somebody would prefer to try to enhance checkpatch to identify only cases where `-ENOSYS` is returned, taking into account that may be from `ret = -ENOSYS; ... ; return ret;` feel free; or decide the status quo is better than changing the message.

Reply:
How about we tweak the text instead, with something along the lines of:
```
Use ENOTSUP instead of ENOSYS as return codes Zephyr APIs. If this warning appears as a result of checking the error code returned by an external API, then this check can be ignored. 
````

Reply:
> How about we tweak the text instead, with something along the lines of:

IMO we should drop this check completely and just document whatever we want in the coding guidelines. This is becoming too verbose and more confusing TBH.

I am going to take my time and think about the whole issue again. I still think it is not the right solution to forbid a return code that has a meaning and usecases just because we made a mistake and we are too deep in the mud. 

We could say that device driver shall use NOTSUP and make it a rule that we do not have to change in the future but still allow usage of NOSYS where it makes sense (outside of drivers).